src/frequenz/sdk/timeseries/_formula_engine/_formula_printer.py

```diff
@@ -0,0 +1,220 @@
+        """
+        match operator:
+            case "+":
+                return OperatorPrecedence.Addition
+            case "-":
+                return OperatorPrecedence.Subtraction
+            case "*":
+                return OperatorPrecedence.Multiplication
+            case "/":
+                return OperatorPrecedence.Division
+            case _:
+                return OperatorPrecedence.Primary
+
+
+class StackItem:
+    """Represents a stack item for the formula printer."""
+
+    def __init__(self, value: str, precedence: OperatorPrecedence):
+        """Initialize the StackItem.
+
+        Args:
+            value: The value of the stack item.
+            precedence: The precedence of the stack item.
+        """
+        self.value = value
+        self.precedence = precedence
+
+    def __str__(self) -> str:
+        """Return the string representation of the stack item.
+
+        This is used for debugging purposes.
+
+        Returns:
+            str: The string representation of the stack item.
+        """
+        return f"(\"{self.value}\", {self.precedence})"
+
+    def plvalue(self, outer_precedence: OperatorPrecedence) -> str:
+        """Return the value of the stack item with parentheses if necessary.
+
+        Args:
+            outer_precedence: The precedence of the outer stack item.
+
+        Returns:
+            str: The value of the stack item with parentheses if necessary.
+        """
+        return f"({self.value})" if self.precedence < outer_precedence else self.value
+
+    def prvalue(self, outer_precedence: OperatorPrecedence) -> str:
+        """Return the value of the stack item with parentheses if necessary.
+
+        Args:
+            outer_precedence: The precedence of the outer stack item.
+
+        Returns:
+            str: The value of the stack item with parentheses if necessary.
+        """
+        return f"({self.value})" if self.precedence < outer_precedence else self.value
+
+    @staticmethod
+    def create_binary(lhs: StackItem, op: str, rhs: StackItem) -> StackItem:
+        """Create a binary stack item.
+
+        Args:
+            lhs: The left-hand side of the binary operation.
+            op: The operator of the binary operation.
+            rhs: The right-hand side of the binary operation.
+
+        Returns:
+            StackItem: The binary stack item.
+        """
+        pred = OperatorPrecedence.map(op)
+        return StackItem(
+            f"{lhs.plvalue(pred)} {op} {rhs.prvalue(pred)}",
+            pred,
+        )
+
+    @staticmethod
+    def create_primary(value: float) -> StackItem:
+        """Create a stack item for literal values or function calls (primary expressions).
+
+        Args:
+            value: The value of the literal.
+
+        Returns:
+            StackItem: The literal stack item.
+        """
+        return StackItem(str(value), OperatorPrecedence.Primary)
+
+
+class FormulaPrinter:
+    """Prints a formula."""
```

Comment:
This is not really printing, it is formatting, so I would name it `FormulaFormatter`.